Add IP range Input support Issue #98

ShubhamRasal opened the pull request.

ShubhamRasal committed:
Add IP range input support (ipv4, ipv6) #98

- Update main.go to accept IP range from stdin and print the possible
  CIDRs

ShubhamRasal committed:
Add test cases to validate RangeTOCIDR function

- Add test case for singleCIDR, multipleCIDRs and ip6 range.
- Convert the []net.IPNet to []string before comparing the expected.

ShubhamRasal committed:
Add check to validate the IP range and Update README

- Accept the list of IP ranges
- Fatal() if the first IP is greater than last IP
- Fatal() if IP range contains more than 2 values.
- Update README.md to add the ip range input documentation.

jdk2588 commented on README.md:
````diff
@@ -183,6 +183,42 @@ In order to list CIDR blocks for given list of IPs, use the following command.
+
+It is also possible to pass the list of IP ranges. The final output will contain the CIDR prefixes of given ranges.
+```
+$ mapcidr -cl 192.168.0.1-192.168.0.255,2c0f:fec9::-2c0f:fed7:ffff:ffff:ffff:ffff:ffff:ffff  -aggregate
````
Do we need to have a list of IP range?

ShubhamRasal replied:
Removed this block from the readme

jdk2588 commented on cmd/mapcidr/main.go:
```diff
@@ -374,12 +387,34 @@ func process(wg *sync.WaitGroup, chancidr, outputchan chan string) {
+			if len(ipRange)%2 == 1 {
+				gologger.Fatal().Msgf("IP range can not have more than 2 values.")
+			}
+			for i := 0; i < len(ipRange)-1; i = i + 2 {
```
This loops is not required, if we need only single IP range?

ShubhamRasal replied:
Removed the for loop

ShubhamRasal committed:
Update IP range functionality to accept only single range

- Remove the unnecessary for loop.
- Update the READ.md to remove the IP range list block

jdk2588 commented on cmd/mapcidr/main.go:
```diff
@@ -393,19 +393,18 @@ func process(wg *sync.WaitGroup, chancidr, outputchan chan string) {
-					outputchan <- cidr.String()
-				}
+			// check if range is valid or not
+			if bytes.Compare(ipRange[0], ipRange[1]) > 0 {
```
Would not the internal function take care of this? Wondering :thinking: , if not then this logic should be in the function.

ShubhamRasal replied:
The internal function doesn't handle any errors. The input validation is handled in main.go.  All the fatal errors are main.go only.

ShubhamRasal replied:
Updated the code and moved the IP comparing and sorting logic to internal function

jdk2588 commented on cmd/mapcidr/main.go:
```diff
@@ -374,12 +387,33 @@ func process(wg *sync.WaitGroup, chancidr, outputchan chan string) {
+			for _, cidr := range cidrs {
+				outputchan <- cidr.String()
+			}
+			options.CIDRFromIPRange = false
```
What is the purpose of setting this to false?

ShubhamRasal replied:
I forgot to remove it. It was just for testing.

ShubhamRasal committed:
Remove the IP comparing logic to a function

- Add new function `GetCIDRFromIPRange` which validate the input
  If input is wrong send the error otherwise returns the sorted list
  of CIDR ranges.
- Update the test case to test `GetCIDRFromIPRange` function.
- Add the wrong input IP range test case.
- Remove the IP comparing and sorting logic from main.go
